---
title: "Protein_Domain_RNA_Localization"
author: "Robert Williams"
date: "3/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# BiocManager::install("biomaRt")
# install.packages("tidyverse")
```


Load Libraries
```{r}
library(biomaRt)
library(tidyverse)
library(openxlsx)
library(ComplexHeatmap)
```

Download information from WormBase ParaSite BioMart.  
Guide is located here: https://parasite.wormbase.org/info/Tools/biomart.html

```{r}
paramart <- useMart("parasite_mart", dataset = "wbps_gene", host = "https://parasite.wormbase.org", port = 443)

protdomain_df <- getBM(
  mart = paramart,
  filter = c("species_id_1010","biotype"),
  values = list(species_id_1010 = "caelegprjna13758", biotype = "protein_coding"),
  attributes = c("production_name_1010", "wormbase_gseq","wbps_gene_id", "wikigene_name", "interpro_id", "interpro_short_description", "interpro_description", "interpro_start", "interpro_end")
)
head(protdomain_df)
```

```{r}
protdomain_df %>% group_by(interpro_description) %>% count() %>% nrow
```
There are 8533 unique protein domain IDs

These are all the protein domains associated with "erm-1", "frm-7", and "imb-2"
```{r}
# protdomain_df %>% filter(wikigene_name == "erm-1")
# protdomain_df %>% filter(wikigene_name == "frm-7")
# protdomain_df %>% filter(interpro_short_description %in% c("PH_domain", "FERM_domain", "PH-like_dom_sf")) %>% group_by(interpro_short_description, interpro_description) %>% count()

goi_domains <- protdomain_df %>% filter(wikigene_name %in% c("erm-1", "frm-7", "imb-2")) %>% group_by(wikigene_name, interpro_short_description, interpro_description) %>% count() %>% arrange(wikigene_name)

goi_domains
```


```{r}
domain_hits <- protdomain_df %>% filter(interpro_short_description %in% goi_domains$interpro_short_description,
  !(interpro_short_description %in% c("Ubiquitin-like_domsf", "ARM-like", "ARM-type_fold"))
  ) %>% 
  select(wbps_gene_id, wikigene_name, interpro_description, interpro_short_description) %>% 
  group_by(wbps_gene_id,wikigene_name, interpro_description, interpro_short_description) %>% 
  count(name = "domain_count") %>%
  ungroup()
domain_hits
length(unique(domain_hits$wbps_gene_id))
```


```{r}
domain_hits_totals <- protdomain_df %>% filter(interpro_short_description %in% goi_domains$interpro_short_description,
  !(interpro_short_description %in% c("Ubiquitin-like_domsf", "ARM-like", "ARM-type_fold"))
  ) %>%  group_by(interpro_short_description, interpro_description) %>% count(name = "domain_count") %>% ungroup() %>% arrange(desc(domain_count)) 
domain_hits_totals
```

```{r}
present_sub <- read.xlsx(xlsxFile ="S1_Dataset_AB_P1_Transcriptome.xlsx",
          sheet = "present_subset",
          startRow = 2) %>% select(WBID)
AB_enr_sub <- read.xlsx(xlsxFile ="S1_Dataset_AB_P1_Transcriptome.xlsx",
          sheet = "AB-enriched_subset",
          startRow = 2) %>% select(WBID)
P1_enr_sub <- read.xlsx(xlsxFile ="S1_Dataset_AB_P1_Transcriptome.xlsx",
          sheet = "P1-enriched_subset",
          startRow = 2) %>% select(WBID)
symm_sub <- read.xlsx(xlsxFile ="S1_Dataset_AB_P1_Transcriptome.xlsx",
          sheet = "symm_subset",
          startRow = 2) %>% select(WBID)
c(nrow(present_sub), nrow(AB_enr_sub), nrow(P1_enr_sub), nrow(symm_sub))
```
Add true/false for different AB/P1 category
```{r}
twocell_domains <- domain_hits %>% 
  mutate(present = case_when(wbps_gene_id %in% present_sub$WBID == TRUE ~ TRUE, 
                                           wbps_gene_id %in% present_sub$WBID == FALSE ~ FALSE),
                       AB_enriched = case_when(wbps_gene_id %in% AB_enr_sub$WBID == TRUE ~ TRUE, 
                                           wbps_gene_id %in% AB_enr_sub$WBID == FALSE ~ FALSE),
                       P1_enriched = case_when(wbps_gene_id %in% P1_enr_sub$WBID == TRUE ~ TRUE, 
                                           wbps_gene_id %in% P1_enr_sub$WBID == FALSE ~ FALSE),
                       symmetric = case_when(wbps_gene_id %in% symm_sub$WBID == TRUE ~ TRUE, 
                                           wbps_gene_id %in% symm_sub$WBID == FALSE ~ FALSE),
                       )
twocell_domains
```



Make upset plot of genes in the `twocell_domains` list for the categories:
"present", "AB_enriched", "P1_enriched" and "symmetric"

```{r}
# remove duplicate gene names which had multiple protein domains
class(twocell_domains)
domain_upset <-
  twocell_domains %>% 
  # ungroup() %>% 
  # filter(present == TRUE) %>%
  select(wikigene_name, present:symmetric) %>% 
  distinct(wikigene_name, .keep_all = TRUE) %>% 
  column_to_rownames(var = "wikigene_name") 
head(domain_upset)
nrow(domain_upset)
sum(domain_upset$present)
sum(domain_upset$AB_enriched)
sum(domain_upset$P1_enriched)
sum(domain_upset$symmetric)
```


```{r}
# replace TRUE/FALSE with 1/0
domain_upset[, sapply(domain_upset, is.logical)] <-
  lapply(domain_upset[, sapply(domain_upset, is.logical)], as.numeric)
# convert to matrix
domain_upset_matrix <- as.matrix.data.frame(domain_upset)
# peek
head(domain_upset_matrix)
# number of rows
nrow(domain_upset_matrix)
```

```{r}
m <- make_comb_mat(domain_upset_matrix)
m
```

Draw the UpSet plot
```{r}
col_size = comb_size(m)
row_size = set_size(m)
ht = UpSet(m, 
	top_annotation = upset_top_annotation(m, ylim = c(0, max(col_size)*1.1)),
	right_annotation = upset_right_annotation(m, ylim = c(0, max(row_size)*1.1)))
ht = draw(ht)

col_od = column_order(ht)
row_od = row_order(ht)
# This stuff adds numbers to the top of the bar plot
decorate_annotation("Intersection\nsize", {
	grid.text(col_size[col_od], 
		seq_len(length(col_size)), 
		unit(col_size[col_od], "native") + unit(2, "mm"), 
		default.units = "native", just = "bottom",
		gp = gpar(fontsize = 8))
})
decorate_annotation("Set size", {
	grid.text(row_size[row_od], 
		unit(row_size[row_od], "native") + unit(2, "mm"), 
		rev(seq_len(length(row_size))), 
		default.units = "native", just = "bottom", rot = -90,
		gp = gpar(fontsize = 8))
})


```

Of the 129 genes present with protein domains similar to erm-1, frm-7 or imb-2:  
- 64 are symmetric
- 6 are AB enriched
- 59 are not significantly differentially expressed
- 20 are not detected in the 2 cell embryo dataset

Additionally, 0 genes are identified to be P1 enriched

Uncomment and run the entire chunk to export the plot
```{r}
# pdf(file = "Protein_Domains_2-Cell_Embryo_Upset_Plot.pdf", width = 5, height = 5)
# ht = draw(ht)
# 
# col_od = column_order(ht)
# row_od = row_order(ht)
# 
# decorate_annotation("Intersection\nsize", {
# 	grid.text(col_size[col_od], 
# 		seq_len(length(col_size)), 
# 		unit(col_size[col_od], "native") + unit(2, "mm"), 
# 		default.units = "native", just = "bottom",
# 		gp = gpar(fontsize = 8))
# })
# decorate_annotation("Set size", {
# 	grid.text(row_size[row_od], 
# 		unit(row_size[row_od], "native") + unit(2, "mm"), 
# 		rev(seq_len(length(row_size))), 
# 		default.units = "native", just = "bottom", rot = -90,
# 		gp = gpar(fontsize = 8))
# })
# dev.off()
```


```{r}
twocell_domains

twocell_domain_genes <- twocell_domains %>% mutate(gene_type = case_when(
  present == TRUE & symmetric == FALSE & AB_enriched == FALSE ~ "no_sig_dif",
  present == TRUE & symmetric == TRUE ~ "symmetric",
  present == TRUE & AB_enriched == TRUE ~ "AB_enriched",
  present == FALSE ~ "not_detected",
)) %>% select(wbps_gene_id:domain_count, gene_type)
twocell_domain_genes

# Number of unique genes in dataset
# Make sure the numbers match the plot above
table((twocell_domain_genes %>% distinct(wikigene_name, .keep_all = TRUE))$gene_type)
```

Number of protein domain types in each two cell embryo gene category
```{r}
table(twocell_domain_genes$gene_type)
```

Get the names of AB enriched genes

```{r}
unique((twocell_domain_genes %>% filter(gene_type == "AB_enriched"))$wikigene_name)
```

Output the data.
```{r}
write.xlsx(twocell_domain_genes, file = "Protein_Domains_2-Cell_Embryo_Genes_210310.xlsx")
```

